src/sync.h

```diff
@@ -70,13 +70,34 @@ inline void EnterCritical(const char* pszName, const char* pszFile, int nLine, v
 inline void LeaveCritical() {}
 inline void CheckLastCritical(void* cs, std::string& lockname, const char* guardname, const char* file, int line) {}
 template <typename MutexType>
-inline void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(cs) {}
+inline void AssertLockHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) {}
 template <typename MutexType>
 void AssertLockNotHeldInternal(const char* pszName, const char* pszFile, int nLine, MutexType* cs) EXCLUSIVE_LOCKS_REQUIRED(!cs) {}
 inline void DeleteLock(void* cs) {}
 inline bool LockStackEmpty() { return true; }
 #endif
-#define AssertLockHeld(cs) AssertLockHeldInternal(#cs, __FILE__, __LINE__, &cs)
+
+/**
+ * Assert at compile time and run time that a mutex is locked. Has no effect
+ * when EXCLUSIVE_LOCKS_REQUIRED annotations are enforced because the runtime
+ * assertions will never trigger (analogous to asserting an unsigned int is
+ * greator or equal to 0).
+ *
+ * Use of this macro is neither encouraged nor discouraged in new code. It was
+ * historically used before compile time checks were available, but it may
+ * still be be useful for virtual methods or std::function callbacks where
+ * EXCLUSIVE_LOCKS_REQUIRED annotations are not enforced.
```

Comment:
> But given that, it makes sense to use `AssertLockHeld` instead of `WeaklyAssertLockHeld` to provide runtime checking inside a lambda annotated with `EXCLUSIVE_LOCKS_REQUIRED` called through a `std::function`

Both the current `LockAssertion` and your proposed `WeaklyAssertLockHeld` do runtime checking, so there's no advantage there.

I don't think AssertLockHeld in lambdas ever makes much sense: there are two ways to use lambdas -- one where thread safety annotations work as expected, and one where they're broken because the lambda's a callback passed to an unannotated function. In the former case you're writing:

```c++
    LOCK(cs);
    [&]() EXCLUSIVE_LOCKS_REQUIRED(cs) { ++foo; }();
```

where it's already obvious that `cs` is held when the lambda is called, and you're just adding the annotation to make the compiler happy.

In the latter case, you're writing:

```c++
    LOCK(cs_main);
    connman->ForEachNode([&](CNode* pnode) {
        WeaklyAssertLockHeld(cs_main);
        ...
    });
```
because you can't pass the fact that `cs_main` is still held through the un-annotated `ForEachNode`. 

> inside an overridden virtual method annotated `EXCLUSIVE_LOCKS_REQUIRED` called through a base method which is not annotated.

I think that is something we should go out of our way to avoid. Either the base method should be annotated as requiring the lock, or the overriding method should acquire the lock itself -- with annotations, locking is part of the function signature, and you shouldn't change the signature when overriding a method. We don't have any code like this now, and I don't think there's any reason ever to, so I'm not sure it's worth addressing it explicitly, but if we are, better to discourage it outright.

Reply:
Thread safety attributes sit on declarations and not on types, but we can check the virtual overriding case. I'll have to look into that. (Theoretically it should be part of the type, but that would interact with other language features like overloading, so it was decided not to put it there.)

Reply:
Looking through the "thread safety" issues on bugzilla, this seems similar to the idea "We could warn in -Wthread-safety-attributes if a declaration has attributes that weren't visible on an earlier declaration." from https://bugs.llvm.org/show_bug.cgi?id=42849